### PR TITLE
Add GPS annotation to the raspivid output

### PR DESCRIPTION
**Pull request by JamesH65**
This involved:
Moving GPS code to own file so can be used in all apps
Adding option in raspivid to apply GPS data to annotation
  Enabled with the -gps flag on command line plus -a 2 to turn on
  application annotation.

### CODE REVIEW

File: host_applications/linux/apps/raspicam/RaspiVid.c [hunk 1]
```diff
@@ -1139,14 +1140,76 @@ static void update_annotation_data(RASPIVID_STATE *state)
...
+         if (gpsdata->set & TIME_SET)
+         {
+            time_t rawtime;
+            struct tm *timeinfo;
+            rawtime = gpsdata->fix.time;
+            timeinfo = localtime(&rawtime);
+            strftime(datetime, sizeof(datetime), "%Y:%m:%d %H:%M:%S", timeinfo);
+         }
+
+
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiVid.c]: Extra blank line.

File: host_applications/linux/apps/raspicam/RaspiVid.c [hunk 2]
```diff
@@ -1139,14 +1140,76 @@ static void update_annotation_data(RASPIVID_STATE *state)
...
+            {
+               if (isnan(gpsdata->fix.latitude) == 0)
+               {
+                  snprintf(lat, sizeof(lat), "%.6lf", gpsdata->fix.latitude);
+               }
+               if (isnan(gpsdata->fix.longitude) == 0)
+               {
+                  snprintf(lon, sizeof(lon), "%.6lf", gpsdata->fix.longitude);
+               }
+            }
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiVid.c]: If I were writing this it would be !isnan and no braces, but as I've been reminded, this isn't the kernel (which means the braces are in the right place).

File: host_applications/linux/apps/raspicam/RaspiVid.c [hunk 3]
```diff
@@ -2408,6 +2471,11 @@ int main(int argc, const char **argv)
 
    check_camera_model(state.common_settings.cameraNum);
 
+   if (state.common_settings.gps)
+   {
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiVid.c]: Braces are used here, but not in the following change. I'd lose them.

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+   gps_reader_data.terminated = 1;
+
+   if (gps_reader_data.gps_reader_thread_ok)
+   {
+      if (verbose)
+         fprintf(stderr, "Waiting for GPS reader thread to terminate\n");
+
+      pthread_join(gps_reader_data.gps_reader_thread, NULL);
+   }
+   if (verbose && (gps_reader_data.gpsd.gpsd_connected))
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiGPS.c]: I see they've been copied from the original but the extra parentheses just make it look more complex than it is.

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+   {
+      int ret = 0;
+      gps_reader_data.gpsd.gpsdata.set = 0;
+      gps_reader_data.gpsd.gpsdata.fix.mode = 0;
+      if ((connect_gpsd(&gps_reader_data.gpsd) < 0) ||
+            ((ret = read_gps_data_once(&gps_reader_data.gpsd)) < 0))
+         break;
+
+      int gps_valid = 0;
+      if ((ret > 0) && (gps_reader_data.gpsd.gpsdata.online))
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiGPS.c]: More extra parens.

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+   while (!gps_reader_data.terminated)
+   {
+      int ret = 0;
+      gps_reader_data.gpsd.gpsdata.set = 0;
+      gps_reader_data.gpsd.gpsdata.fix.mode = 0;
+      if ((connect_gpsd(&gps_reader_data.gpsd) < 0) ||
+            ((ret = read_gps_data_once(&gps_reader_data.gpsd)) < 0))
+         break;
+
+      int gps_valid = 0;
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiGPS.c]: We're not supposed to use C++ style declarations...

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+      gps_reader_data.terminated = 1;
+      raspi_gps_shutdown(verbose);
+      exit(EX_SOFTWARE);
+      return;
+   }
+
+   gps_reader_data.gps_reader_thread_ok = 1;
+   return;
+}
+
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiGPS.c]: KILL THE TRAILING BLANK LINE!!!11!??

File: host_applications/linux/apps/raspicam/RaspiGPS.c (hunk 6 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiGPS.c]: This was a C&P from the previous code in Raspistill, which came from a third party, so was relucatant tochange anything in it, but will do anyway!

File: host_applications/linux/apps/raspicam/RaspiGPS.c (hunk 7 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiGPS.c]: Arghghghghghg

File: host_applications/linux/apps/raspicam/RaspiVid.c (hunk 2 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiVid.c]: Have removed some but not all, for reasons of readability rather than consistency.

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      fprintf(stderr, "Closing gpsd connection\n\n");
+
+   disconnect_gpsd(&gps_reader_data.gpsd);
+
+   libgps_unload(&gps_reader_data.gpsd);
+
+   pthread_mutex_destroy(&gps_reader_data.gps_cache_mutex);
+}
+
+
```
Comment (pelwell) [inline host_applications/linux/apps/raspicam/RaspiGPS.c]: The other example of extra whitespace.

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+{
+   memset(&gps_reader_data, 0, sizeof(gps_reader_data));
+
+   pthread_mutex_init(&gps_reader_data.gps_cache_mutex, NULL);
+
+   gpsd_init(&gps_reader_data.gpsd);
+   if (libgps_load(&gps_reader_data.gpsd))
+   {
+      pthread_mutex_destroy(&gps_reader_data.gps_cache_mutex);
+      exit(EX_SOFTWARE);
```
Comment (6by9) [inline host_applications/linux/apps/raspicam/RaspiGPS.c]: exit here seems a tad extreme. Return error and exit from the caller (if that is deemed necessary). Ditto the other failure cases in this function.

File: host_applications/linux/apps/raspicam/RaspiVid.c [hunk 10]
```diff
@@ -542,13 +544,13 @@ static void application_help_message(char *app_name)
       fprintf(stdout, ",%s", raw_output_fmt_map[i].mode);
    }
 
-   fprintf(stdout, "\n");
+   fprintf(stdout, "\n\n");
```
Comment (6by9) [inline host_applications/linux/apps/raspicam/RaspiVid.c]: Unrelated change ;-)

File: host_applications/linux/apps/raspicam/RaspiVid.c [hunk 11]
```diff
@@ -1139,14 +1140,72 @@ static void update_annotation_data(RASPIVID_STATE *state)
...
+                  (!isnan(gpsdata->fix.altitude)))
+            {
+               snprintf(alt, sizeof(alt), "Altitude=%.2fm",
+                        (gpsdata->fix.altitude*10+0.5)/10.0 );
+            }
+
+            if ((gpsdata->set & SPEED_SET) && !isnan(gpsdata->fix.speed))
+            {
+               snprintf(speed, sizeof(speed), "Speed=%.2fms",
+                        ((gpsdata->fix.speed*MPS_TO_KPH*10+0.5)/10) );
```
Comment (6by9) [inline host_applications/linux/apps/raspicam/RaspiVid.c]: What units are we using where?
Text says milliseconds(!!!), and we're taking a speed (gps.h says in m/s), multiplying by MPS_TO_KPH, and rounding. So the text should say km/h?

File: host_applications/linux/apps/raspicam/CMakeLists.txt [hunk 12]
```diff
@@ -43,9 +43,11 @@ set (COMMON_SOURCES
    RaspiCLI.c
    RaspiPreview.c
    RaspiCommonSettings.c
-   RaspiHelpers.c)
+   RaspiHelpers.c
+   RaspiGPS.c
+   libgps_loader.c)
```
Comment (6by9) [inline host_applications/linux/apps/raspicam/CMakeLists.txt]: You've added it to COMMON_SOURCES, so raspividyuv and raspiyuv will now accept the parameter but do nothing with them.

Comment (JamesH65) [reply host_applications/linux/apps/raspicam/CMakeLists.txt]: Yes, I thought the benfits of making it common for two of the apps outweighed the fact its not (yet) used in the YUV applications.

File: host_applications/linux/apps/raspicam/RaspiVid.c (hunk 11 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiVid.c]: Hmm. Need to double check that. I thoughts it was m/s, but will cross check against the cgps test app.

File: host_applications/linux/apps/raspicam/RaspiVid.c (hunk 10 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiVid.c]: Yep....noticed a formating problem while testing. I coudl send it in a different PR, but I have a horrible feeling the life is too short.

File: host_applications/linux/apps/raspicam/RaspiGPS.c (hunk 9 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiGPS.c]: Hmm, the original dropped out, not nececessarily so abruptly. Will ponder how to make this neater.

File: host_applications/linux/apps/raspicam/RaspiGPS.c [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+   pthread_mutex_lock(&gps_reader_data.gps_cache_mutex);
+   return &gps_reader_data.gpsdata_cache;
+}
+
+void raspi_gps_unlock()
+{
+   pthread_mutex_unlock(&gps_reader_data.gps_cache_mutex);
+}
+
+
```
[the remark this answers is not in the document]
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiGPS.c]: I really need to adjust my `astyle` parameters to fix that automatically.

File: host_applications/linux/apps/raspicam/RaspiGPS.c (hunk 9 above)
Comment (6by9) [reply host_applications/linux/apps/raspicam/RaspiGPS.c]: Cleanup (not needed here as gpsd_init only does a memset and writes a couple of values) and return an error code.
The caller can abort on error, but it makes this a more useful library than having it randomly abort the app.

The code looks to be doing all the cleanup already, so simply replacing the ```exit``` calls with return -1, and return 0 at the end would do it. (Function needs to change from returning void to returning int too).

File: host_applications/linux/apps/raspicam/RaspiVid.c (hunk 10 above)
Comment (6by9) [reply host_applications/linux/apps/raspicam/RaspiVid.c]: :-)

File: host_applications/linux/apps/raspicam/RaspiVid.c (hunk 11 above)
Comment (6by9) [reply host_applications/linux/apps/raspicam/RaspiVid.c]: gps.h says 
```    double speed;	/* Speed over ground, meters/sec */```
so I think it wants to be either 
```"Speed=%.2fkm/h", ((gpsdata->fix.speed*MPS_TO_KPH*10+0.5)/10) ```
or
```"Speed=%.2fm/s", gpsdata->fix.speed) ```

File: host_applications/linux/apps/raspicam/CMakeLists.txt (hunk 12 above)
Comment (6by9) [reply host_applications/linux/apps/raspicam/CMakeLists.txt]: I'll let you off then.

Comment (JamesH65) [reply host_applications/linux/apps/raspicam/CMakeLists.txt]: Phew

File: host_applications/linux/apps/raspicam/RaspiVid.c (hunk 11 above)
Comment (JamesH65) [reply host_applications/linux/apps/raspicam/RaspiVid.c]: Yup, a number of those calcs can be much simplified. Another C&P thingy.

**Comment by JamesH65** [comment]
OK, so this latest push is a little more than just the cleanups suggested above. I have moved the location string generation to the RaspGPS file so that it can now be used in all the apps and have modified all the apps to use it. So you can now get the GPS annotation on stills and video, in normal or yuv/rgb form.

**Comment by JamesH65** [comment]
Thanks all. Except for the extra line at the end comment. Grr.
